### PR TITLE
feat(cards): add support for MS CryptoAPI tokens

### PR DESCRIPTION
**Pull request by mrts**
WE2-716

### CODE REVIEW

**Comment by mrts** [comment]
Change the [allowed PIN symbols regex](https://github.com/web-eid/web-eid-app/blob/main/src/ui/webeiddialog.cpp#L479-L481) to add support for the character ranges defined in the [SafeNet Authentication Client Administrator Guide](https://installer.id.ee/media/etoken/Linux_SAC%2010.8.28%20GA%20Build.zip):

> The Administrator password quality and Initialization Key quality must include three out of the following four
rules:
> 1. English uppercase letters (ASCII 0x41...0x5A)
> 2. English lowercase letters (ASCII 0x61...0x7A)
> 3. Numeric (ASCII 0x30...0x39)
> 4. Special characters (ASCII 0x20...0x2F + 0x3A...0x40 + 0x5B...0x60 + 0x7B...0x7F)

Add support for UTF letters (umlauts etc) as well.

**Review by mrts — COMMENTED** [review]
Change the [allowed PIN symbols regex](https://github.com/web-eid/web-eid-app/blob/main/src/ui/webeiddialog.cpp#L479-L481) to add support for the character ranges defined in the [SafeNet Authentication Client Administrator Guide](https://installer.id.ee/media/etoken/Linux_SAC%2010.8.28%20GA%20Build.zip):

> The Administrator password quality and Initialization Key quality must include three out of the following four
rules:
> 1. English uppercase letters (ASCII 0x41...0x5A)
> 2. English lowercase letters (ASCII 0x61...0x7A)
> 3. Numeric (ASCII 0x30...0x39)
> 4. Special characters (ASCII 0x20...0x2F + 0x3A...0x40 + 0x5B...0x60 + 0x7B...0x7F)

Add support for UTF letters (umlauts etc) as well.

File: src/controller/command-handlers/certificatereader.cpp [hunk 1]
```diff
@@ -85,7 +85,11 @@ void CertificateReader::run(const std::vector<CardInfo::ptr>& cards)
     std::vector<CardCertificateAndPinInfo> certInfos;
     certInfos.reserve(cards.size());
     for (const auto& card : cards) {
-        certInfos.emplace_back(getCertificateWithStatusAndInfo(card, certificateType));
+        try {
+            certInfos.emplace_back(getCertificateWithStatusAndInfo(card, certificateType));
```
Comment (metsma) [inline src/controller/command-handlers/certificatereader.cpp:89]:
```suggestion
            certInfos.push_back(getCertificateWithStatusAndInfo(card, certificateType));
```

File: src/ui/webeiddialog.cpp [hunk 2]
```diff
@@ -458,6 +469,31 @@ void WebEidDialog::setTrText(QWidget* label, const std::function<QString()>& tex
...
+        // This is required as accessing widgets from background threads is not allowed,
+        // so getPin() cannot access pinInput directly.
+        // QString uses QAtomicPointer internally and is thread-safe.
+        pin = ui->pinInput->text();
+
+        // TODO: We need to erase the PIN in the widget buffer, this needs further work.
+        // Investigate if it is possible to keep the PIN in secure memory, e.g. with a
+        // custom Qt widget.
+        // Clear the PIN input.
+        ui->pinInput->setText({});
```
Comment (metsma) [inline src/ui/webeiddialog.cpp:491]:
```suggestion
        ui->pinInput->clear();
```

File: src/controller/command-handlers/certificatereader.cpp (hunk 1 above)
Comment (mrts) [reply src/controller/command-handlers/certificatereader.cpp:89]: Thanks, fixed.

File: src/ui/webeiddialog.cpp (hunk 2 above)
Comment (mrts) [reply src/ui/webeiddialog.cpp:491]: Thanks, fixed.

**Pull request merged**